src/main/scala/is/hail/expr/ir/Simplify.scala

```diff
@@ -71,12 +71,31 @@ object Simplify {
         
       case Let(n, v, b) if !Mentions(b, n) => b
 
+      case ArrayFold(ArrayMap(a, n1, b), zero, accumName, valueName, body) => ArrayFold(a, zero, accumName, n1, Let(valueName, b, body))
```

Comment:
the arraymap (and any other array ops within a) will get deforested inside the ArrayFold anyway. I think the two sides are basically equivalent.

Reply:
I agree, however, I think Let is going to expose additional optimization opportunities downstream so it should be preferred.